json/decode.go

```diff
@@ -564,6 +564,18 @@ func (d decoder) decodeArray(b []byte, p unsafe.Pointer, n int, size uintptr, t
 	}
 }
 
+//go:linkname unsafe_NewArray reflect.unsafe_NewArray
```

Comment:
This is going to have consequences when this unexported name changes. We'll have to advertise which versions of the go stdlib work with our package, and we'll get bug reports about the confusing compilation failure.

Reply:
The API has existed for at least 7 years https://github.com/golang/go/blame/4e8d27068df52eb372dc2ba7e929e47850934805/src/reflect/value.go#L2249-L2250
This would be pretty much documenting that we support Go until the latest 🤷‍♂ 

Here's what the error message will look like if Go ever removes this API:
```
# github.com/segmentio/encoding/json.test
github.com/segmentio/encoding/json.decoder.decodeSlice: relocation target reflect.unsafe_NewArray2 not defined
FAIL	github.com/segmentio/encoding/json [build failed]
FAIL
```
This seems easy enough to catch, but I may just be biased because I'm too familiar with the problem space.

Reply:
The logical conclusion of that argument is that we should feel free to use external, unexported functions whenever we like as long as they don't seem to change often. It kinda defeats the purpose of differentiating public API from private implementation.

Suppose we depend on (say) the Go 1.19 version of the reflect package, because in Go 1.19 they changed the behavior of this unexported function -- a totally reasonable thing for them to do.

Say another package does something similar, but they depend on the Go 1.18 behavior. It's now impossible for anyone to use both packages at the same time. Modules requirements won't help them here.

Reply:
Yes, you make very good points.

I guess my experience with other open-source projects (as a user or maintainer) has taught me that documenting compatibility promises does little in preventing issues of this kind, I think it also goes against the design goals of this project to not add extra responsibilities on the users.

Here's something we could try:
- add a `json/reflect_optimize.go` file which with build tags to compile against Go 1.13 and below
- add a `json/reflect.go` file which with build tags to compile against Go 1.14+, and relies on the slower but public APIs of the `reflect` package.

For each new Go release that still has the unexported APIs we rely on here, we will update the build tags to carry the optimization forward, but until we do the compiler will select the less optimized version in `json/reflect.go`. That way the code is always correct, and a bit faster on Go releases that we have validated to support the optimizations.

This does put extra work on the maintainers, but it's the trade off we chose to make with this project, so it fits better with the original design decisions.

What do you think?

Reply:
Is there also the option of `+build go1.13` tags to specify which file to use for a Go version?

Reply:
Ah, I didn't consider build tags. That seems OK to me. Honestly maybe you could just use `+build ignore` or a custom one like `+build safe` for the portable version, for reference, until the day arrives that the build breaks and we need to support a different `reflect` implementation. If one wants to test against the portable version, or maybe somebody doesn't want to run with that in production (unlikely if they are using this package), it's easy to enable it via the `-tags` build flag.

Reply:
if you prefer the strategy of updating the tag for each release once it's proven to work, that's absolutely fine with me too

Reply:
I think I like the option of supporting newer versions on a less optimized but correct code paths, it means programs using tip still work with the package, it keeps overhead on the users as low as possible.